src/WebRTCSession.cpp

```diff
@@ -891,51 +908,106 @@ WebRTCSession::addVideoPipeline(int vp8PayloadType)
+                gst_caps_unref(caps);
+
+                gst_bin_add_many(GST_BIN(pipe_), camera, camerafilter, nullptr);
+                if (!gst_element_link_many(camera, videoconvert, camerafilter, nullptr)) {
+                        nhlog::ui()->error("WebRTC: failed to link camera elements");
+                        return false;
+                }
+                if (callType_ == CallType::VIDEO && !gst_element_link(camerafilter, tee)) {
+                        nhlog::ui()->error("WebRTC: failed to link camerafilter -> tee");
                         return false;
                 }
-                g_object_set(source, "use-damage", FALSE, nullptr);
-                g_object_set(source, "xid", 0, nullptr);
-                auto settings = ChatPage::instance()->userSettings();
-                g_object_set(source, "show-pointer", !settings->screenShareHideCursor(), nullptr);
+        }
+
+        if (callType_ == CallType::SCREEN) {
+                nhlog::ui()->debug("WebRTC: screen share frame rate: {} fps",
+                                   settings->screenShareFrameRate());
+                nhlog::ui()->debug("WebRTC: screen share picture-in-picture: {}",
+                                   settings->screenSharePiP());
+                nhlog::ui()->debug("WebRTC: screen share request remote camera: {}",
+                                   settings->screenShareRemoteVideo());
                 nhlog::ui()->debug("WebRTC: screen share hide mouse cursor: {}",
                                    settings->screenShareHideCursor());
-                int frameRate = settings->screenShareFrameRate();
-                caps          = gst_caps_new_simple(
-                  "video/x-raw", "framerate", GST_TYPE_FRACTION, frameRate, 1, nullptr);
-                nhlog::ui()->debug("WebRTC: screen share frame rate: {} fps", frameRate);
-        }
 
-        GstElement *videoconvert = gst_element_factory_make("videoconvert", nullptr);
-        GstElement *capsfilter   = gst_element_factory_make("capsfilter", "camerafilter");
-        g_object_set(capsfilter, "caps", caps, nullptr);
-        gst_caps_unref(caps);
+                GstElement *ximagesrc = gst_element_factory_make("ximagesrc", "screenshare");
+                if (!ximagesrc) {
+                        nhlog::ui()->error("WebRTC: failed to create ximagesrc");
+                        return false;
+                }
+                g_object_set(ximagesrc, "use-damage", FALSE, nullptr);
+                g_object_set(ximagesrc, "xid", 0, nullptr);
+                g_object_set(
+                  ximagesrc, "show-pointer", !settings->screenShareHideCursor(), nullptr);
+
+                GstCaps *caps          = gst_caps_new_simple("video/x-raw",
+                                                    "framerate",
+                                                    GST_TYPE_FRACTION,
+                                                    settings->screenShareFrameRate(),
```

Comment:
Not something for this PR, but could this be made dynamic in the future, if the connection quality drops or so?

Reply:
I'd have to research that tbh. For now setting a low frame rate seems to work ok unless you're sharing video say.